econml/validate/drtester.py

```diff
@@ -0,0 +1,796 @@
+            self.dr_train = self.calculate_dr_outcomes(Dtrain, ytrain, reg_preds_train, prop_preds_train)
+
+            # Get DR outcomes in validation sample
+            reg_preds_val, prop_preds_val = self.fit_nuisance_train(Xtrain, Dtrain, ytrain, Xval)
+            self.dr_val = self.calculate_dr_outcomes(Dval, yval, reg_preds_val, prop_preds_val)
+
+            self.Dtrain = Dtrain
+
+        else:
+            # Get DR outcomes in validation sample
+            reg_preds_val, prop_preds_val = self.fit_nuisance_cv(Xval, Dval, yval)
+            self.dr_val = self.calculate_dr_outcomes(Dval, yval, reg_preds_val, prop_preds_val)
+
+        # Calculate ATE in the validation sample
+        self.ate_val = self.dr_val.mean(axis=0)
+
+        return self
+
+    def fit_cate(
+        self,
+        reg_cate,
+        Zval: np.array,
+        Ztrain: np.array = None
+    ):
+
+        """
+        Fits CATE model and generates predictions. If Ztrain is None, then the predictions are generated using k-folds
+        cross-validation on the validation set. If Ztrain is specified, then the CATE is fit on the training sample
+        (where the DR outcomes were generated using k-folds CV), and then applied to the validation sample.
+
+        Parameters
+        ----------
+        reg_cate: estimator
+            CATE model. Must be able to implement `fit' and `predict' methods
+        Zval: (n_val x n_treatment) matrix
+            Validation set features to be used to predict (and potentially fit) DR outcomes in CATE model
+        Ztrain (n_train x n_treatment) matrix, defaul ``None``
+            Training set features used to fit CATE model
+
+        Returns
+        -------
+        None, but adds attribute cate_preds_val for predicted CATE values on the validation set and, if training
+        data is provided, attribute cate_preds_train for predicted CATE values on the training set
+
+        """
+
+        if (Ztrain is None) and self.fit_on_train:
+            raise Exception("Nuisance models fit on training sample but Ztrain not specified")
+
+        if (Ztrain is not None) and (not self.fit_on_train):
+            raise Exception("Nuisance models cross-fit on validation sample but Ztrain is specified")
+
+        if Ztrain is not None:
+            self.cate_preds_train = self.fit_cate_cv(reg_cate, Ztrain, self.Dtrain, self.dr_train, self.n_splits)
+            self.cate_preds_val = self.fit_cate_train(reg_cate, Ztrain, Zval)
+        else:
+            self.cate_preds_val = self.fit_cate_cv(reg_cate, Zval, self.Dval, self.dr_val, self.n_splits)
+
+    def evaluate_cal(
+        self,
+        reg_cate=None,
+        Zval: np.array = None,
+        Ztrain: np.array = None,
+        n_groups: int = 4
+    ):
+
+        """
+        Implements calibration test as in [Dwivedi2020]
+
+        Parameters
+        ----------
+        reg_cate: estimator, default ``None``
```

Comment:
The cate estimator should be adhering to the EconML CATE API. This is not fit and predict. Rather it is fit and effect(T=1), effect(T=2), const_marginal_effect etc. 

Also here you should not be using `fit` you should be assuming the CATE estimator is already fitted. So just specify the methods that are required to be supported by the fitted CATE estimator. I suspect this would just be `cate.effect(T=t)`, where t are the different treatment values sorted in lexicographic order.

Reply:
Make sure you can handle the case when the input treatments are for instance alphanumeric, i.e. I pass a T of the form ['a', 'b', 'a', 'a', 'b', etc]